### PR TITLE
implement sniffing of datatypes for lddas

### PR DESCRIPTION
**Pull request by martenson**
implements https://github.com/galaxyproject/galaxy/issues/7336

### CODE REVIEW

**Comment by martenson** [comment]
the https://github.com/galaxyproject/galaxy/pull/7379/commits/3f64fef118939155cff92007d73059d2e5554c23 commit maybe justifies backport

File: lib/galaxy/managers/datasets.py [hunk 1]
```diff
@@ -335,6 +337,37 @@ def serialize_dataset_association_roles(self, trans, dataset_assoc):
             rval["modify_item_roles"] = modify_item_role_list
         return rval
 
+    def detect_datatype(self, trans, dataset_assoc):
+        """Sniff and assign the datatype to a given dataset associsation (ldda or hda)"""
```
Comment (nsoranzo) [inline lib/galaxy/managers/datasets.py]:
```suggestion
        """Sniff and assign the datatype to a given dataset association (ldda or hda)"""
```

**Review by nsoranzo — COMMENTED** [review]
Is the failed API test relevant? https://jenkins.galaxyproject.org/job/docker-api/14609/testReport/junit/test.api.test_libraries/LibrariesApiTestCase/test_update_dataset_in_folder/

**Comment by martenson** [comment]
@nsoranzo I think so, will check it out tomorrow

**Comment by martenson** [comment]
With ebd8c44 in place Galaxy is creating jobs in db without a session_id and that seems to work. I think this means that anonymous set_metadata jobs will forever be orphaned, since they won't have user or session_id. Which is probably fine (@natefoo ?). Thanks @jmchilton.

**Comment by natefoo** [comment]
> this means that anonymous set_metadata jobs will forever be orphaned, since they won't have user or session_id

This doesn't seem great - I feel like I'm missing some understanding here though?

**Comment by jmchilton** [comment]
Tools API controller:

```
        history_id = payload.get('history_id', None)
        if history_id:
            decoded_id = self.decode_id(history_id)
            target_history = self.history_manager.get_owned(decoded_id, trans.user, current_history=trans.history)
        else:
            target_history = None
```

Tool actions:

```
        if hasattr(trans, "get_galaxy_session"):
            galaxy_session = trans.get_galaxy_session()
            # If we're submitting from the API, there won't be a session.
            if type(galaxy_session) == trans.model.GalaxySession:
                job.session_id = model.cached_id(galaxy_session)
        if trans.user is not None:
            job.user_id = model.cached_id(trans.user)
        job.history_id = model.cached_id(history)
        job.tool_id = tool.id
```

Clearly these things can be null already, there is no problem here IMO.

**Comment by martenson** [comment]
the last commit was wrongly cherry picked here and is also part of https://github.com/galaxyproject/galaxy/pull/7397, it should create no conflict I think

File: lib/galaxy/tools/actions/metadata.py [hunk 2]
```diff
@@ -18,8 +18,11 @@ def execute(self, tool, trans, incoming={}, set_output_hid=False, overwrite=True
         Execute using a web transaction.
         """
         trans.check_user_activation()
-        job, odict = self.execute_via_app(tool, trans.app, trans.get_galaxy_session().id,
-                                          trans.history.id, trans.user, incoming, set_output_hid,
+        session = trans.get_galaxy_session()
+        session_id = session and session.id
```
Comment (mvdbeek) [inline lib/galaxy/tools/actions/metadata.py:22]: If this is supposed to be a boolean `has_session_id` might be a better variable name, but I think this should probably be `session_id = session.id if session else None` or `session_id = getattr(session, 'id', None)`?

File: lib/galaxy/managers/library_datasets.py [hunk 3]
```diff
@@ -115,7 +121,7 @@ def _validate_and_parse_update_payload(self, payload):
                 validated_payload[key] = val
             if key in ('file_ext'):
                 datatype = self.app.datatypes_registry.get_datatype_by_extension(val)
-                if datatype is None:
+                if datatype is None and val not in ("auto"):
```
Comment (nsoranzo) [inline lib/galaxy/managers/library_datasets.py]:
```suggestion
                if datatype is None and val not in ("auto",):
```

File: lib/galaxy/tools/actions/metadata.py (hunk 2 above)
Comment (nsoranzo) [reply lib/galaxy/tools/actions/metadata.py:22]: I think this already equivalent to `session_id = session.id if session else None`

File: lib/galaxy/managers/library_datasets.py (hunk 3 above)
Comment (martenson) [reply lib/galaxy/managers/library_datasets.py]: How did this pass lint? Thanks.

File: lib/galaxy/tools/actions/metadata.py (hunk 2 above)
Comment (martenson) [reply lib/galaxy/tools/actions/metadata.py:22]: @nsoranzo that is my understanding also, I like @mvdbeek's explicit proposal though

File: lib/galaxy/managers/library_datasets.py (hunk 3 above)
Comment (nsoranzo) [reply lib/galaxy/managers/library_datasets.py]: Because it's equivalent to `val not in "auto"` :
```
In [5]: "auto" in "auto"
Out[5]: True
```

Comment (martenson) [reply lib/galaxy/managers/library_datasets.py]: Python tryin to be helpful, nice.

File: lib/galaxy/tools/actions/metadata.py (hunk 2 above)
Comment (mvdbeek) [reply lib/galaxy/tools/actions/metadata.py:22]: Learning something new every day
```python
>>> 'a' and 'b'
'b'
```
I would have expected `True` here.

**Comment by martenson** [comment]
@mvdbeek I've added a basic API test that triggers the datatype detection. It is fairly primitive, I could add something that actually checks the eventually sniffed datatype but that seemed out of scope for an API test. We test sniffing elsewhere.

Edit: when put like this the test actually does not test much... I guess I will add the check for sniffed extension.

**Comment by martenson** [comment]
@mvdbeek I upgraded the test, now it actually tests the sniffing (for one datatype only, just the process)

**Review by mvdbeek — APPROVED** [review]
Thanks @martenson, looks good to me.

**Pull request merged**